src/Compilers/CSharp/Test/Semantic/Semantics/GenericConstraintsTests.cs

```diff
@@ -3270,5 +3264,209 @@ void Test()
                 //         UnmanagedWithInterface(&a);         // fail (does not match interface)
                 Diagnostic(ErrorCode.ERR_GenericConstraintNotSatisfiedValType, "UnmanagedWithInterface").WithArguments("C.UnmanagedWithInterface<T>(T*)", "System.IDisposable", "T", "int").WithLocation(20, 9));
         }
+
+        [Fact]
+        public void UnmanagedGenericStructPointer()
+        {
+            var code = @"
+public struct MyStruct<T>
+{
+    public T field;
+}
+
+public class C
+{
+    public unsafe void M()
+    {
+        MyStruct<int> myStruct;
+        M2(&myStruct);
+    }
+
+    public unsafe void M2(MyStruct<int> *ms) { }
+}
+";
+            CreateCompilation(code, options: TestOptions.UnsafeReleaseDll)
+                .VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ManagedGenericStructPointer()
+        {
+            var code = @"
+public struct MyStruct<T>
+{
+    public T field;
+}
+
+public class C
+{
+    public unsafe void M()
+    {
+        MyStruct<string> myStruct;
+        M2(&myStruct);
+    }
+
+    public unsafe void M2<T>(MyStruct<T>* ms) where T : unmanaged { }
+}
+";
+            CreateCompilation(code, options: TestOptions.UnsafeReleaseDll)
+                .VerifyDiagnostics(
+                    // (12,12): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('MyStruct<string>')
+                    //         M2(&myStruct);
+                    Diagnostic(ErrorCode.ERR_ManagedAddr, "&myStruct").WithArguments("MyStruct<string>").WithLocation(12, 12));
+        }
+
+        [Fact]
+        public void UnmanagedGenericConstraintStructPointer()
```

Comment:
Should have a test for the T type parameter case constrained to unmanaged where the type is instantiated (normally not as a pointer) where the type argument is:

- a reference type
- a struct that contains a struct that has only unmanaged types
- a generic struct that is instantiated with a unmanaged type that makes it unmanaged:
- a generic struct that is instantiated with a type parameter that makes it ambiguous (and hence doesn't meet the constraint)

Appologies if they are in this PR and I missed them.

Reply:
> a generic struct that is instantiated with a type parameter that makes it ambiguous (and hence doesn't meet the constraint)

Can you give an example for this? Does this refer to partial construction?